precompiles/crowdloan-rewards/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+	Runtime::Call: From<pallet_crowdloan_rewards::Call<Runtime>>,
+{
+	// The accessors are first. They directly return their result.
+	fn is_contributor(input: &[u8]) -> Result<PrecompileOutput, ExitError> {
+		// parse the address
+		let contributor = H160::from_slice(&input[12..32]);
+
+		log::trace!(
+			target: "crowdloan-rewards-precompile",
+			"Checking whether {:?} is a contributor",
+			contributor
+		);
+
+		let account: Runtime::AccountId = contributor.into();
+		// fetch data from pallet
+		let is_contributor =
+			pallet_crowdloan_rewards::Pallet::<Runtime>::accounts_payable(account).is_some();
+
+		log::trace!(target: "crowldoan-rewards-precompile", "Result from pallet is {:?}", is_contributor);
+
+		let gas_consumed = <Runtime as pallet_evm::Config>::GasWeightMapping::weight_to_gas(
+			<Runtime as frame_system::Config>::DbWeight::get().read,
+		);
+
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gas_consumed,
+			output: bool_to_solidity_bytes(is_contributor),
+			logs: Default::default(),
+		})
+	}
+
+	fn reward_info(input: &[u8]) -> Result<PrecompileOutput, ExitError> {
+		// parse the address
+		let contributor = H160::from_slice(&input[12..32]);
+
+		log::trace!(
+			target: "crowdloan-rewards-precompile",
+			"Checking reward info for {:?}",
+			contributor
+		);
+
+		let account: Runtime::AccountId = contributor.into();
+		// fetch data from pallet
+		let reward_info = pallet_crowdloan_rewards::Pallet::<Runtime>::accounts_payable(account);
+
+		let (total, claimed): (U256, U256) = if let Some(reward_info) = reward_info {
+			let total_reward: u128 = reward_info.total_reward.try_into().map_err(|_| {
+				ExitError::Other("Amount is too large for provided balance type".into())
+			})?;
+			let claimed_reward: u128 = reward_info.claimed_reward.try_into().map_err(|_| {
+				ExitError::Other("Amount is too large for provided balance type".into())
+			})?;
+
+			(total_reward.into(), claimed_reward.into())
+		} else {
+			(0u128.into(), 0u128.into())
+		};
+
+		log::trace!(
+			target: "crowldoan-rewards-precompile", "Result from pallet is {:?}  {:?}",
+			total, claimed
+		);
+
+		let gas_consumed = <Runtime as pallet_evm::Config>::GasWeightMapping::weight_to_gas(
+			<Runtime as frame_system::Config>::DbWeight::get().read,
+		);
+		let mut buffer = [0u8; 64];
+		total.to_big_endian(&mut buffer[0..32]);
+		claimed.to_big_endian(&mut buffer[32..64]);
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gas_consumed,
+			output: buffer.to_vec(),
+			logs: Default::default(),
+		})
+	}
+
+	fn claim() -> Result<pallet_crowdloan_rewards::Call<Runtime>, ExitError> {
+		Ok(pallet_crowdloan_rewards::Call::<Runtime>::claim())
+	}
+}
+
+// Solidity's bool type is 256 bits as shown by these examples
+// https://docs.soliditylang.org/en/v0.8.0/abi-spec.html
+// This utility function converts a Rust bool into the corresponding Solidity type
+fn bool_to_solidity_bytes(b: bool) -> Vec<u8> {
```

Comment:
We should really DRY this (I think it's copied here from `parachain-staking`)...

Reply:
indeed

Reply:
I am just thinking of making a precompiles-common package that we can import here